File: mobile/android/app/src/main/java/com/anonymous/focusshell/AppBlockerModule.java
package com.anonymous.focusshell;

import android.app.ActivityManager;
import android.app.usage.UsageStats;
import android.app.usage.UsageStatsManager;
import android.content.Context;
import android.content.Intent;
import android.provider.Settings;
import android.util.Log;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.Arguments;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.WritableMap;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.SortedMap;
import java.util.TreeMap;

/**
 * AppBlockerModule - Block and track app usage
 * 
 * Provides:
 * - Block specific apps
 * - Unblock apps
 * - Check if app is blocked
 * - Get foreground app
 * - Request usage stats permission
 */
public class AppBlockerModule extends ReactContextBaseJavaModule {
    private static final String TAG = "AppBlockerModule";
    private final ReactApplicationContext reactContext;
    private final Set<String> blockedApps = new HashSet<>();

    public AppBlockerModule(ReactApplicationContext context) {
        super(context);
        this.reactContext = context;
    }

    @NonNull
    @Override
    public String getName() {
        return "AppBlocker";
    }

    /**
     * Block a list of apps
     */
    @ReactMethod
    public void blockApps(ReadableArray packageNames, Promise promise) {
        try {
            for (int i = 0; i < packageNames.size(); i++) {
                String packageName = packageNames.getString(i);
                if (packageName != null && !packageName.isEmpty()) {
                    blockedApps.add(packageName);
                    Log.d(TAG, "Blocked app: " + packageName);
                }
            }
            
            WritableMap result = Arguments.createMap();
            result.putBoolean("success", true);
            result.putInt("blockedCount", packageNames.size());
            
            promise.resolve(result);
        } catch (Exception e) {
            Log.e(TAG, "Error blocking apps", e);
            promise.reject("ERROR", e.getMessage());
        }
    }

    /**
     * Unblock a list of apps
     */
    @ReactMethod
    public void unblockApps(ReadableArray packageNames, Promise promise) {
        try {
            for (int i = 0; i < packageNames.size(); i++) {
                String packageName = packageNames.getString(i);
                if (packageName != null) {
                    blockedApps.remove(packageName);
                    Log.d(TAG, "Unblocked app: " + packageName);
                }
            }
            
            WritableMap result = Arguments.createMap();
            result.putBoolean("success", true);
            result.putInt("unblockedCount", packageNames.size());
            
            promise.resolve(result);
        } catch (Exception e) {
            Log.e(TAG, "Error unblocking apps", e);
            promise.reject("ERROR", e.getMessage());
        }
    }

    /**
     * Unblock all apps
     */
    @ReactMethod
    public void unblockAllApps(Promise promise) {
        try {
            int count = blockedApps.size();
            blockedApps.clear();
            
            WritableMap result = Arguments.createMap();
            result.putBoolean("success", true);
            result.putInt("unblockedCount", count);
            
            promise.resolve(result);
        } catch (Exception e) {
            Log.e(TAG, "Error unblocking all apps", e);
            promise.reject("ERROR", e.getMessage());
        }
    }

    /**
     * Check if an app is blocked
     */
    @ReactMethod
    public void isAppBlocked(String packageName, Promise promise) {
        try {
            boolean isBlocked = blockedApps.contains(packageName);
            
            WritableMap result = Arguments.createMap();
            result.putBoolean("isBlocked", isBlocked);
            result.putString("packageName", packageName);
            
            promise.resolve(result);
        } catch (Exception e) {
            Log.e(TAG, "Error checking if app is blocked", e);
            promise.reject("ERROR", e.getMessage());
        }
    }

    /**
     * Get the currently foreground app
     */
    @ReactMethod
    public void getForegroundApp(Promise promise) {
        try {
            String foregroundApp = getCurrentForegroundApp();
            
            WritableMap result = Arguments.createMap();
            result.putString("packageName", foregroundApp != null ? foregroundApp : "");
            result.putBoolean("isBlocked", foregroundApp != null && blockedApps.contains(foregroundApp));
            
            promise.resolve(result);
        } catch (Exception e) {
            Log.e(TAG, "Error getting foreground app", e);
            promise.reject("ERROR", e.getMessage());
        }
    }

    /**
     * Check if app has usage stats permission
     */
    @ReactMethod
    public void hasUsageStatsPermission(Promise promise) {
        try {
            UsageStatsManager usageStatsManager = (UsageStatsManager) reactContext
                .getSystemService(Context.USAGE_STATS_SERVICE);
            
            long currentTime = System.currentTimeMillis();
            List<UsageStats> stats = usageStatsManager.queryUsageStats(
                UsageStatsManager.INTERVAL_DAILY,
                currentTime - 1000 * 60,
                currentTime
            );
            
            boolean hasPermission = stats != null && !stats.isEmpty();
            
            WritableMap result = Arguments.createMap();
            result.putBoolean("hasPermission", hasPermission);
            
            promise.resolve(result);
        } catch (Exception e) {
            WritableMap result = Arguments.createMap();
            result.putBoolean("hasPermission", false);
            promise.resolve(result);
        }
    }

    /**
     * Request usage stats permission
     */
    @ReactMethod
    public void requestUsageStatsPermission(Promise promise) {
        try {
            Intent intent = new Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS);
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            reactContext.startActivity(intent);
            
            WritableMap result = Arguments.createMap();
            result.putBoolean("success", true);
            result.putString("message", "Usage stats settings opened");
            
            promise.resolve(result);
        } catch (Exception e) {
            Log.e(TAG, "Error requesting usage stats permission", e);
            promise.reject("ERROR", e.getMessage());
        }
    }

    /**
     * Bring Zen Mobile launcher to foreground (to block app)
     */
    @ReactMethod
    public void bringLauncherToForeground(Promise promise) {
        try {
            Intent intent = reactContext.getPackageManager()
                .getLaunchIntentForPackage(reactContext.getPackageName());
            
            if (intent != null) {
                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TOP);
                reactContext.startActivity(intent);
                
                WritableMap result = Arguments.createMap();
                result.putBoolean("success", true);
                
                promise.resolve(result);
            } else {
                promise.reject("ERROR", "Could not create launcher intent");
            }
        } catch (Exception e) {
            Log.e(TAG, "Error bringing launcher to foreground", e);
            promise.reject("ERROR", e.getMessage());
        }
    }

    /**
     * Get current foreground app package name
     */
    private String getCurrentForegroundApp() {
        try {
            UsageStatsManager usageStatsManager = (UsageStatsManager) reactContext
                .getSystemService(Context.USAGE_STATS_SERVICE);
            
            long currentTime = System.currentTimeMillis();
            
            // Query usage stats for last 1 minute
            List<UsageStats> stats = usageStatsManager.queryUsageStats(
                UsageStatsManager.INTERVAL_DAILY,
                currentTime - 1000 * 60,
                currentTime
            );
            
            if (stats != null && !stats.isEmpty()) {
                SortedMap<Long, UsageStats> sortedStats = new TreeMap<>();
                for (UsageStats usageStats : stats) {
                    sortedStats.put(usageStats.getLastTimeUsed(), usageStats);
                }
                
                if (!sortedStats.isEmpty()) {
                    return sortedStats.get(sortedStats.lastKey()).getPackageName();
                }
            }
        } catch (Exception e) {
            Log.w(TAG, "Error getting foreground app", e);
        }
        
        return null;
    }

    /**
     * Check if blocked app is being launched and intercept
     */
    public boolean shouldBlockApp(String packageName) {
        return blockedApps.contains(packageName);
    }
}

File: mobile/android/app/src/main/java/com/anonymous/focusshell/modules/AppBlockerModule.kt
package com.anonymous.focusshell.modules

import android.app.AppOpsManager
import android.app.usage.UsageStats
import android.app.usage.UsageStatsManager
import android.content.Context
import android.content.Intent
import android.os.Build
import android.provider.Settings
import com.facebook.react.bridge.*
import java.util.*

class AppBlockerModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private val blockedApps = mutableSetOf<String>()

    override fun getName(): String {
        return "AppBlocker"
    }

    /**
     * Block apps during focus sessions
     */
    @ReactMethod
    fun blockApps(packageNames: ReadableArray, promise: Promise) {
        try {
            var blockedCount = 0
            for (i in 0 until packageNames.size()) {
                val packageName = packageNames.getString(i)
                if (packageName != null) {
                    blockedApps.add(packageName)
                    blockedCount++
                }
            }

            val result = WritableNativeMap()
            result.putBoolean("success", true)
            result.putInt("blockedCount", blockedCount)
            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("ERROR", "Failed to block apps: ${e.message}")
        }
    }

    /**
     * Unblock apps
     */
    @ReactMethod
    fun unblockApps(packageNames: ReadableArray, promise: Promise) {
        try {
            var unblockedCount = 0
            for (i in 0 until packageNames.size()) {
                val packageName = packageNames.getString(i)
                if (packageName != null && blockedApps.remove(packageName)) {
                    unblockedCount++
                }
            }

            val result = WritableNativeMap()
            result.putBoolean("success", true)
            result.putInt("unblockedCount", unblockedCount)
            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("ERROR", "Failed to unblock apps: ${e.message}")
        }
    }

    /**
     * Unblock all apps
     */
    @ReactMethod
    fun unblockAllApps(promise: Promise) {
        try {
            val count = blockedApps.size
            blockedApps.clear()

            val result = WritableNativeMap()
            result.putBoolean("success", true)
            result.putInt("unblockedCount", count)
            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("ERROR", "Failed to unblock all apps: ${e.message}")
        }
    }

    /**
     * Check if app is blocked
     */
    @ReactMethod
    fun isAppBlocked(packageName: String, promise: Promise) {
        try {
            val result = WritableNativeMap()
            result.putBoolean("isBlocked", blockedApps.contains(packageName))
            result.putString("packageName", packageName)
            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("ERROR", "Failed to check if app is blocked: ${e.message}")
        }
    }

    /**
     * Get foreground app
     */
    @ReactMethod
    fun getForegroundApp(promise: Promise) {
        try {
            // This requires usage stats permission
            if (!hasUsageStatsPermissionInternal()) {
                val result = WritableNativeMap()
                result.putString("packageName", "unknown")
                result.putBoolean("isBlocked", false)
                promise.resolve(result)
                return
            }

            val usageStatsManager = reactApplicationContext
                .getSystemService(Context.USAGE_STATS_SERVICE) as UsageStatsManager

            val time = System.currentTimeMillis()
            val stats = usageStatsManager.queryUsageStats(
                UsageStatsManager.INTERVAL_DAILY,
                time - 1000 * 10,
                time
            )

            var foregroundApp = "unknown"
            if (stats != null && stats.isNotEmpty()) {
                val sortedStats = stats.sortedByDescending { it.lastTimeUsed }
                foregroundApp = sortedStats.firstOrNull()?.packageName ?: "unknown"
            }

            val result = WritableNativeMap()
            result.putString("packageName", foregroundApp)
            result.putBoolean("isBlocked", blockedApps.contains(foregroundApp))
            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("ERROR", "Failed to get foreground app: ${e.message}")
        }
    }

    /**
     * Check if has usage stats permission
     */
    @ReactMethod
    fun hasUsageStatsPermission(promise: Promise) {
        try {
            val hasPermission = hasUsageStatsPermissionInternal()
            val result = WritableNativeMap()
            result.putBoolean("hasPermission", hasPermission)
            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("ERROR", "Failed to check usage stats permission: ${e.message}")
        }
    }

    /**
     * Request usage stats permission
     */
    @ReactMethod
    fun requestUsageStatsPermission(promise: Promise) {
        try {
            val intent = Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
            reactApplicationContext.startActivity(intent)

            val result = WritableNativeMap()
            result.putBoolean("success", true)
            result.putString("message", "Usage stats settings opened")
            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("ERROR", "Failed to open usage stats settings: ${e.message}")
        }
    }

    /**
     * Bring launcher to foreground
     */
    @ReactMethod
    fun bringLauncherToForeground(promise: Promise) {
        try {
            val intent = reactApplicationContext.packageManager
                .getLaunchIntentForPackage(reactApplicationContext.packageName)
            
            if (intent != null) {
                intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TOP
                reactApplicationContext.startActivity(intent)
                
                val result = WritableNativeMap()
                result.putBoolean("success", true)
                result.putString("message", "Launcher brought to foreground")
                promise.resolve(result)
            } else {
                promise.reject("ERROR", "Failed to get launcher intent")
            }
        } catch (e: Exception) {
            promise.reject("ERROR", "Failed to bring launcher to foreground: ${e.message}")
        }
    }

    private fun hasUsageStatsPermissionInternal(): Boolean {
        return try {
            val appOps = reactApplicationContext
                .getSystemService(Context.APP_OPS_SERVICE) as AppOpsManager
            
            val mode = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                appOps.unsafeCheckOpNoThrow(
                    AppOpsManager.OPSTR_GET_USAGE_STATS,
                    android.os.Process.myUid(),
                    reactApplicationContext.packageName
                )
            } else {
                @Suppress("DEPRECATION")
                appOps.checkOpNoThrow(
                    AppOpsManager.OPSTR_GET_USAGE_STATS,
                    android.os.Process.myUid(),
                    reactApplicationContext.packageName
                )
            }
            
            mode == AppOpsManager.MODE_ALLOWED
        } catch (e: Exception) {
            false
        }
    }
}
